kollect-core/src/main/kotlin/kollect/Kollect.kt

```diff
@@ -30,20 +34,129 @@ sealed class KollectQuery<I : Any, A> : KollectRequest() {
  * Primitive operations in the Kollect Free monad.
  */
 @higherkind
-interface KollectOp<A> : TC
+sealed class KollectOp<out A> : KollectOpOf<A> {
+    data class Thrown<A>(val err: Throwable) : KollectOp<A>()
+    data class Join<A, B>(val fl: Kollect<A>, val fr: Kollect<B>) : KollectOp<Tuple2<A, B>>()
+    data class Concurrent(val queries: NonEmptyList<KollectQuery<Any, Any>>) : KollectOp<InMemoryCache>(), KollectRequest
+    data class KollectOne<I : Any, A>(val id: I, val ds: DataSource<I, A>) : KollectOp<A>(), KollectQuery<I, A> {
+        override fun dataSource(): DataSource<I, A> = ds
+        override fun identities(): NonEmptyList<I> = NonEmptyList.pure(id)
+    }
+
+    data class KollectMany<I : Any, A>(val ids: NonEmptyList<I>, val ds: DataSource<I, A>) : KollectOp<List<A>>(), KollectQuery<I, A> {
+        override fun dataSource(): DataSource<I, A> = ds
+        override fun identities(): NonEmptyList<I> = ids
+    }
+
+    companion object {
+
+        /**
+         * Lift a plain value to the Kollect monad.
+         */
+        fun <A> pure(a: A): Kollect<A> = Free.pure(a)
+
+        /**
+         * Lift an exception to the Kollect monad.
+         */
+        fun <A> error(e: Throwable): Free<ForKollectOp, A> = Free.liftF(KollectOp.Thrown(e))
+
+        /**
+         * Given a value that has a related `DataSource` implementation, lift it
+         * to the `Kollect` monad. When executing the kollect the data source will be
+         * queried and the kollect will return its result.
+         */
+        fun <I : Any, A> apply(ds: DataSource<I, A>, i: I): Kollect<A> =
+                Free.liftF(KollectOp.KollectOne(i, ds))
+
+        /**
+         * Given multiple values with a related `DataSource` lift them to the `Kollect` monad.
+         */
+        fun <I : Any, A> multiple(ds: DataSource<I, A>, i: I, vararg ids: I): Kollect<List<A>> =
+                Free.liftF(KollectOp.KollectMany(NonEmptyList(i, ids.toList()), ds))
+
+        /**
+         * Transform a list of kollects into a kollect of a list. It implies concurrent execution of kollects.
+         */
+        fun <I : Any, A> sequence(ids: List<Kollect<A>>): Kollect<List<A>> = traverse(ids, { x -> x })
+
+        /**
+         * Apply a kollect-returning function to every element in a list and return a Kollect of the list of
+         * results. It implies concurrent execution of kollects.
+         */
+        fun <A, B> traverse(ids: List<A>, f: (A) -> Kollect<B>): Kollect<List<B>> =
+                traverseGrouped(ids, 50, f)
+
+        fun <A> ListK<A>.grouped(n: Int): ListK<ListK<A>> =
+            withIndex()
+                    .groupBy { it.index / n }
+                    .map { it.value.k().map { it.value } }.k()
```

Comment:
Apparently my previous comment got lost somewhere.

I am not sure if I understand the last line 100%, but I think this `grouped` implementation will result in a `traverse` that returns the results in a different order than the input.
It seems to me that `(1,2,3,4,5)` with `grouped(2)` would become `((1,3,5), (2,4))`?

Reply:
I can't assure that it's the best implementation but it works for me. You can try it by yourself at [try.kotlinlang.org](https://try.kotlinlang.org/#/UserProjects/str02igkvqj1tevjid3pb0d421/42bcgmd91f4b673helqqvbv6et) 

```kotlin
fun main(args: Array<String>) {
    fun <A> List<A>.grouped(n: Int): List<List<A>> =
            withIndex()
                    .groupBy { it.index / n }
                    .map { it.value.map { it.value } }

	val a = listOf(1,2,3,4,5)

	println(a.grouped(2))
}
```

Reply:
Yes, you are right, `it.index % n` would do what I mentioned before.
We can always come back and improve this implementation.